packages/api/lambdas/sf-scheduler.js

```diff
@@ -3,64 +3,65 @@
 const get = require('lodash.get');
 const merge = require('lodash.merge');
 const uuidv4 = require('uuid/v4');
-const { getS3Object, parseS3Uri } = require('@cumulus/common/aws');
+const { getMessageFromTemplate } = require('@cumulus/ingest/queue');
 const { SQS } = require('@cumulus/ingest/aws');
 const { Provider, Collection } = require('../models');
 
-/**
- * Builds a cumulus-compatible message and adds it to the startSF queue
- * startSF queue will then start a stepfunction for the given message
- *
- * @param   {Object} event   - lambda input message
- * @param   {Object} context - lambda context
- * @param   {function} cb    - lambda callback
- */
-function schedule(event, context, cb) {
-  const template = get(event, 'template');
+const buildCumulusMeta = (queueName) => ({
+  execution_name: uuidv4(),
+  queueName
+});
+
+function getProvider(providerId) {
+  if (providerId) {
+    const p = new Provider();
+    return p.get({ id: providerId });
+  }
+  return {};
+}
+
+function getCollection(collection) {
+  if (collection) {
+    const c = new Collection();
+    return c.get({ name: collection.name, version: collection.version });
+  }
+  return {};
+}
+
+function buildMessage(event, baseMessage) {
   const provider = get(event, 'provider', null);
   const meta = get(event, 'meta', {});
   const cumulusMeta = get(event, 'cumulus_meta', {});
   const collection = get(event, 'collection', null);
   const payload = get(event, 'payload', {});
-  let message;
+  const queueName = get(event, 'queueName', 'startSF');
+
+  return {
+    ...baseMessage,
+    provider: getProvider(provider),
+    collection: getCollection(collection),
+    meta: merge(baseMessage.meta, meta),
+    payload,
+    cumulus_meta: merge(cumulusMeta, buildCumulusMeta(queueName))
+  };
+}
+
+/**
+ * Add a Cumulus workflow message to the queue specified by event.queueName.
+ * 
+ * A consumer should be configured for this queue to start executions for
+ * the queued message.
+ *
+ * @param {Object} event - lambda input message
+ */
+async function schedule(event) {
+  const template = get(event, 'template');
 
-  const parsed = parseS3Uri(template);
-  getS3Object(parsed.Bucket, parsed.Key)
-    .then((data) => {
-      message = JSON.parse(data.Body);
-      message.meta.provider = {};
-      message.meta.collection = {};
-      message.meta = merge(message.meta, meta);
-      message.payload = payload;
-      message.cumulus_meta.execution_name = uuidv4();
-      message.cumulus_meta = merge(message.cumulus_meta, cumulusMeta);
-    })
-    .then(() => {
-      if (provider) {
-        const p = new Provider();
-        return p.get({ id: provider });
-      }
-      return null;
-    })
-    .then((p) => {
-      if (p) {
-        message.meta.provider = p;
-      }
+  const data = await getMessageFromTemplate(template);
```

Comment:
Can we rename `data` to `baseMessage` or `messageTemplate`?

Reply:
Good idea. I like `messageTemplate` more than `baseMessage`... I might have to propagate that through my other functions, too.